### PR TITLE
executor: evaluate subquery before executing outer query.

### PR DESCRIPTION
**Pull request by coocood**
So outer query can push down subquery values.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @hanfei19910905 PTAL

**Comment by hanfei19910905** [comment]
LGTM

File: ast/expressions.go [hunk 1]
```diff
@@ -265,26 +263,14 @@ func (n *SubqueryExpr) Accept(v Visitor) (Node, bool) {
 		return v.Leave(newNode)
 	}
 	n = newNode.(*SubqueryExpr)
-
 	if n.SubqueryExec != nil {
-		t, ok := n.SubqueryExec.Accept(v)
```
Comment (shenli) [inline ast/expressions.go]: Why not visit SubqueryExec?

File: executor/adapter.go [hunk 2]
```diff
@@ -100,3 +109,65 @@ func (a *statement) Exec(ctx context.Context) (ast.RecordSet, error) {
...
+}
+
+func (e *subqueryExecutor) Leave(in ast.Node) (ast.Node, bool) {
+	switch x := in.(type) {
+	case *ast.CompareSubqueryExpr:
+		e.multipleRows = false
+	case *ast.PatternInExpr:
+		e.multipleRows = false
+	case *ast.ExistsSubqueryExpr:
+		e.existRow = false
```
Comment (shenli) [inline executor/adapter.go]: Why set this back?

File: executor/adapter.go [hunk 3]
```diff
@@ -100,3 +109,65 @@ func (a *statement) Exec(ctx context.Context) (ast.RecordSet, error) {
...
+	case *ast.SubqueryExpr:
+		if !x.UseOuterContext && x.SubqueryExec != nil {
+			rowCount := 2
```
Comment (shenli) [inline executor/adapter.go]: Seems this is duplicate with Evaluator.subqueryExpr().

File: executor/compiler.go [hunk 4]
```diff
@@ -50,6 +50,7 @@ func (c *Compiler) Compile(ctx context.Context, node ast.StmtNode) (ast.Statemen
 	sa := &statement{
 		is:   is,
 		plan: p,
+		node: node,
```
Comment (shenli) [inline executor/compiler.go]: Any comment for this?

File: ast/expressions.go (hunk 1 above)
Comment (coocood) [inline ast/expressions.go]: `SubqueryExec` is generated object, it should not be an `ast.Node`.

File: executor/adapter.go (hunk 2 above)
Comment (coocood) [inline executor/adapter.go]: There may be multiple subqueries.

File: executor/adapter.go (hunk 3 above)
Comment (coocood) [inline executor/adapter.go]: It's looks similar but is different.

File: executor/compiler.go (hunk 4 above)
Comment (coocood) [inline executor/compiler.go]: node will be used to evalutate subquery.

File: executor/adapter.go (hunk 3 above)
Comment (coocood) [inline executor/adapter.go]: `Evaluator` can not be used directly.
I prefer  `copy on write`, so one doesn't affect the other.

**Comment by coocood** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**